### PR TITLE
Draw text using decimal numbers

### LINKED ISSUE #972: Bad letter spacing with canvas
Depending on the font and font size, the letters are spaced inaccurately. For example, using the `monaco` font with a font size of `12px` will result in a real character width of `7.203...px` but charMeasure does `Match.ceil` it to `8px`. This creates a visible `0.8px` additional gap between characters, which causes the font to look somewhat stretched:

Current (with `Math.ceil`):
wide

Expected (without `Math.ceil`):
correct

@Tyriar Do we have to Math.ceil the measured dimensions for some reason?

### PR DESCRIPTION
**Pull request by Tyriar**
Fixes #972

### CODE REVIEW

**Comment by Tyriar** [comment]
Looks a lot better now. The top one is after the change, both windows are using the same font/zoom level:

screen shot 2017-09-12 at 7 07 46 pm

File: src/renderer/BaseRenderLayer.ts [hunk 1]
```diff
@@ -79,14 +79,29 @@ export abstract class BaseRenderLayer implements IRenderLayer {
...
    * @param width The number of columns to fill.
    * @param height The number of rows to fill.
    */
   protected fillCells(x: number, y: number, width: number, height: number): void {
-    this._ctx.fillRect(x * this.scaledCharWidth, y * this.scaledLineHeight, width * this.scaledCharWidth, height * this.scaledLineHeight);
+    const cellLeft = this._getCellLeft(x);
+    this._ctx.fillRect(
+      cellLeft,
+      y * this.scaledLineHeight,
+      this._getCellLeft(x + width) - cellLeft,
```
Comment (mofux) [inline src/renderer/BaseRenderLayer.ts]: How about adding a `this._getCellWidth(x, cells)` function that does does this calculation for us? Would make the code a little easier to read.

File: src/renderer/BaseRenderLayer.ts [hunk 2]
```diff
@@ -96,10 +111,11 @@ export abstract class BaseRenderLayer implements IRenderLayer {
    * @param y The row to fill.
    */
   protected fillBottomLineAtCells(x: number, y: number, width: number = 1): void {
+    const cellLeft = this._getCellLeft(x);
     this._ctx.fillRect(
-        x * this.scaledCharWidth,
+        cellLeft,
         (y + 1) * this.scaledLineHeight - window.devicePixelRatio - 1 /* Ensure it's drawn within the cell */,
-        width * this.scaledCharWidth,
+        this._getCellLeft(x + width) - cellLeft,
```
Comment (mofux) [inline src/renderer/BaseRenderLayer.ts]: Could re-use `this._getCellWidth(x, cells)`

File: src/renderer/BaseRenderLayer.ts [hunk 3]
```diff
@@ -124,11 +140,12 @@ export abstract class BaseRenderLayer implements IRenderLayer {
...
    */
   protected strokeRectAtCell(x: number, y: number, width: number, height: number): void {
+    const cellLeft = this._getCellLeft(x);
     this._ctx.lineWidth = window.devicePixelRatio;
     this._ctx.strokeRect(
-        x * this.scaledCharWidth + window.devicePixelRatio / 2,
+        cellLeft + window.devicePixelRatio / 2,
         y * this.scaledLineHeight + (window.devicePixelRatio / 2),
-        (width * this.scaledCharWidth) - window.devicePixelRatio,
+        this._getCellLeft(x + width) - cellLeft - window.devicePixelRatio,
```
Comment (mofux) [inline src/renderer/BaseRenderLayer.ts]: Could re-use `this._getCellWidth(x, cells)`

File: src/renderer/BaseRenderLayer.ts [hunk 4]
```diff
@@ -147,7 +164,12 @@ export abstract class BaseRenderLayer implements IRenderLayer {
    * @param height The number of rows to clear.
    */
   protected clearCells(x: number, y: number, width: number, height: number): void {
-    this._ctx.clearRect(x * this.scaledCharWidth, y * this.scaledLineHeight, width * this.scaledCharWidth, height * this.scaledLineHeight);
+    const cellLeft = this._getCellLeft(x);
+    this._ctx.clearRect(
+        cellLeft,
+        y * this.scaledLineHeight,
+        this._getCellLeft(x + width) - cellLeft,
```
Comment (mofux) [inline src/renderer/BaseRenderLayer.ts]: Could re-use `this._getCellWidth(x, cells)`

File: src/renderer/BaseRenderLayer.ts (hunk 1 above)
Comment (Tyriar) [reply src/renderer/BaseRenderLayer.ts]: Sounds good :smiley:

Comment (mofux) [reply src/renderer/BaseRenderLayer.ts]: Sounds reasonable 👍 Guess we're good to go then!

**Pull request merged**